src/ray/core_worker/object_interface.cc

```diff
@@ -35,32 +55,26 @@ Status CoreWorkerObjectInterface::Get(const std::vector<ObjectID> &ids,
                                       std::vector<std::shared_ptr<RayObject>> *results) {
   (*results).resize(ids.size(), nullptr);
 
-  // Divide the object ids into two groups: direct call return objects and the rest,
-  // and de-duplicate for each group.
-  std::unordered_set<ObjectID> direct_call_return_ids;
-  std::unordered_set<ObjectID> other_ids;
-  for (const auto &object_id : ids) {
-    if (object_id.IsReturnObject() &&
-        object_id.GetTransportType() ==
-            static_cast<int>(TaskTransportType::DIRECT_ACTOR)) {
-      direct_call_return_ids.insert(object_id);
-    } else {
-      other_ids.insert(object_id);
-    }
-  }
+  // There can be a few cases here:
+  // - for task return objects, find the store provider type for an object from
+  //   its transport, and then try to get from the corresponding store provider;
+  // - for other objects, try to get from plasma store.
+  EnumUnorderedMap<StoreProviderType, std::unordered_set<ObjectID>>
+      object_ids_per_store_provider;
+  GetObjectIdsPerStoreProvider(ids, &object_ids_per_store_provider);
 
   std::unordered_map<ObjectID, std::shared_ptr<RayObject>> objects;
-  auto start_time = current_time_ms();
-  // Fetch non-direct-call objects using `PLASMA` store provider.
-  RAY_RETURN_NOT_OK(Get(StoreProviderType::PLASMA, other_ids, timeout_ms, &objects));
-  int64_t duration = current_time_ms() - start_time;
-  int64_t left_timeout_ms =
-      (timeout_ms == -1) ? timeout_ms
-                         : std::max(static_cast<int64_t>(0), timeout_ms - duration);
-
-  // Fetch direct call return objects using `LOCAL_PLASMA` store provider.
-  RAY_RETURN_NOT_OK(Get(StoreProviderType::LOCAL_PLASMA, direct_call_return_ids,
-                        left_timeout_ms, &objects));
+  auto current_timeout_ms = timeout_ms;
+
+  for (const auto &entry : object_ids_per_store_provider) {
```

Comment:
Document that if getting from one store provider uses up the timeout, we will still try the others with a timeout of 0.

Reply:
sure